File: base/src/main/java/de/hipp/pnp/Attribute5e.java
package de.hipp.pnp;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class Attribute5e {
    @JsonIgnore
    Integer baseValue = 0;

    Integer value = 0;
    Integer max = 20;
    Integer modifier = 0;

    public Attribute5e(Integer baseValue) {
        this.baseValue = baseValue;
        this.value = baseValue;
    }

    public Attribute5e(Integer baseValue, Integer max) {
        this.baseValue = baseValue;
        this.value = baseValue;
        this.max = max;
    }

    public Integer getModifier() {
        return (value - 10) / 2;
    }

    public void modifyValue(Integer value) {
        this.value = Math.min(max, baseValue + value);
    }
}

File: base/src/main/kotlin/de/hipp/pnp/base/fivee/Attribute5e.kt
package de.hipp.pnp.base.fivee

import com.fasterxml.jackson.annotation.JsonIgnore
import java.io.Serializable
import kotlin.math.min

/**
 * Represents a 5th Edition attribute with value, maximum, and modifier.
 *
 * @property baseValue The base value of the attribute (default: 0)
 * @property value The current value of the attribute
 * @property max The maximum value the attribute can reach (default: 20)
 * @property modifier The calculated modifier based on the attribute value
 */
data class Attribute5e(
    @JsonIgnore var baseValue: Int = 0,
) : Serializable {
    var value: Int = baseValue

    @JvmField
    var max: Int = 20
    var modifier: Int = 0

    /**
     * Modifies the attribute value and recalculates the modifier.
     * The value is capped at the maximum allowed value.
     *
     * @param value The amount to add to the current value
     */
    fun modifyValue(value: Int) {
        this.value = min(max, this.value + value)
        this.modifier = (this.value - 10).floorDiv(2)
    }
}
